web/app/components/input/input.tsx

```diff
@@ -148,10 +172,163 @@ export class Input extends Component<Props, State> {
+    const maxImageSizeString = (StaticStore.config.max_image_size / 1024 / 1024).toFixed(2) + 'MB';
+    /** upload delay to avoid server rate limiter */
+    const uploadDelay = 5000;
+
+    const isSelectionSupported = this.textAreaRef.isSelectionSupported();
+
+    this.setState({
+      errorLock: true,
+      errorMessage: null,
+      isErrorShown: false,
+      isDisabled: true,
+      buttonText: 'Uploading...',
+    });
+
+    // fallback for ie < 9
+    if (!isSelectionSupported) {
+      for (let i = 0; i < data.length; i++) {
+        const file = data[i];
+        const isFirst = i === 0;
+        const placeholderStart = this.state.text.length === 0 ? '' : '\n';
+
+        if (file.size > StaticStore.config.max_image_size) {
+          this.appendError(`${file.name} exceeds size limit of ${maxImageSizeString}`);
+          continue;
+        }
+
+        !isFirst && (await sleep(uploadDelay));
+
+        const result = await this.props
+          .uploadImage(file)
+          .catch(
+            (e: ApiError | string) =>
+              new Error(
+                typeof e === 'string'
+                  ? `${file.name} upload failed with "${e}"`
+                  : `${file.name} upload failed with "${e.error}"`
+              )
+          );
+
+        if (result instanceof Error) {
+          this.appendError(result.message);
+          continue;
+        }
+
+        const markdownString = `${placeholderStart}![${result.name}](${result.url})`;
+        this.setState({
+          text: this.state.text + markdownString,
+        });
+      }
+
+      this.setState({ errorLock: false, isDisabled: false, buttonText: null });
+      return;
+    }
+
+    for (let i = 0; i < data.length; i++) {
+      const file = data[i];
+      const isFirst = i === 0;
+      const placeholderStart = this.state.text.length === 0 ? '' : '\n';
+
+      const uploadPlaceholder = `${placeholderStart}![uploading ${file.name}...]()`;
+      const uploadPlaceholderLength = uploadPlaceholder.length;
+      const selection = this.textAreaRef.getSelection();
+      /** saved selection in case of error */
+      const originalText = this.state.text;
+      const restoreSelection = async () => {
+        this.setState({
+          text: originalText,
+        });
+        /** sleeping awhile so textarea catch state change and its selection */
+        await sleep(100);
+        this.textAreaRef!.setSelection(selection);
+      };
+
+      if (file.size > StaticStore.config.max_image_size) {
+        this.appendError(`${file.name} exceeds size limit of ${maxImageSizeString}`);
+        continue;
+      }
+
+      this.setState({
+        text: replaceSelection(this.state.text, selection, uploadPlaceholder),
+      });
+
+      !isFirst && (await sleep(uploadDelay));
+
+      const result = await this.props
+        .uploadImage(file)
+        .catch(
+          (e: ApiError | string) =>
+            new Error(
+              typeof e === 'string'
```

Comment:
Code duplication

Reply:
removed code duplication, also noticed that `this.textAreaRef.isSelectionSupported` returned `false` if selectionStart is 0, ouch.